packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -0,0 +1,461 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../aux/access/ISubmitBlocksCallback.sol";
+import "../aux/transactions/BlockReader.sol";
+import "../thirdparty/BytesUtil.sol";
+import "../core/iface/IExchangeV3.sol";
+import "../lib/AddressUtil.sol";
+import "../lib/ERC20.sol";
+import "../lib/ERC20SafeTransfer.sol";
+import "../lib/MathUint.sol";
+
+import "../core/impl/libtransactions/AmmUpdateTransaction.sol";
+import "../core/impl/libtransactions/DepositTransaction.sol";
+import "../core/impl/libtransactions/WithdrawTransaction.sol";
+
+contract AmmPool is ISubmitBlocksCallback {
+
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using BlockReader       for ExchangeData.Block;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event Deposit(
+        address  owner,
+        uint96[] amounts
+    );
+
+    event Withdrawal(
+        address  owner,
+        uint96[] amounts
+    );
+
+    event JoinPoolRequested(
+        address  owner,
+        uint     poolAmountOut,
+        uint96[] maxAmountsIn
+    );
+
+    event ExitPoolRequested(
+        address  owner,
+        uint     poolAmountIn,
+        uint96[] minAmountsOut
+    );
+
+    uint public constant MAX_AGE_REQUEST_UNTIL_POOL_SHUTDOWN = 7 days;
+
+    IExchangeV3 public exchange;
+    uint32      public accountID;
+
+    uint        public shutdownTimestamp = 0;
+
+    struct PoolJoin
+    {
+        uint poolAmountOut;
```

Comment:
The amount of liquidity tokens the LP expects to receive/spend. Used what balancer does here to check that we can do what we need to do, but I haven't actually implemented LP tokens here because hopefully we won't have to do that ourselves.

Reply:
I would suggest to remove them or rename them to something like "ownershipTokenBurnt/Mint"

Reply:
Uniswap calls them "liquidity tokens", Balancer/Bancor calls them "pool tokens". Would stick with one of them so we don't name things differently.